### PR TITLE
add stakingGet/Put/DelDelegate API

### PR DESCRIPTION
**Pull request by dustinxie**

### CODE REVIEW

File: action/protocol/staking/delegate.go [hunk 1]
```diff
@@ -247,7 +269,55 @@ func (m DelegateMap) Deserialize(buf []byte) error {
 	}
 
 	for _, v := range pb.Delegates {
-		m[hash.BytesToHash160(v.Owner)] = fromProto(v)
+		m[v.Owner] = fromProto(v)
 	}
 	return nil
 }
+
+func stakingGetDelegate(sr protocol.StateReader, name address.Address) (*Delegate, error) {
```
Comment (CoderZhi) [inline action/protocol/staking/delegate.go]: why we need to add a prefix "staking" to the function name?

File: action/protocol/staking/stakingpb/staking.proto [hunk 2]
```diff
@@ -20,7 +20,7 @@ message BucketIndices {
 }
 
 message Delegate {
-    bytes owner = 1;
+    string owner = 1;
     string address = 2;
     string rewardAddress = 3;
```
Comment (CoderZhi) [inline action/protocol/staking/stakingpb/staking.proto]: nit: to be space efficient, it is ok to store owner, address, rewardadress as bytes.

File: action/protocol/staking/delegate.go [hunk 3]
```diff
@@ -7,16 +7,28 @@
...
+)
+
+/*
+ * keys in namespace factory.DelegateNameSpace are prefixed with 1-byte tag
+ * to indicate the type of stored object. This enables to quickly scan the
+ * entire bucket and filter out objects of a certain type
+ */
+const (
+	inActiveDel = byte(iota)
+	activeDel   = inActiveDel + 1
```
Comment (CoderZhi) [inline action/protocol/staking/delegate.go]: could you elaborate or comment on the usage of these two parameters?

File: action/protocol/staking/delegate.go [hunk 4]
```diff
@@ -7,16 +7,28 @@
...
+	"github.com/iotexproject/iotex-core/state/factory"
+)
+
+/*
+ * keys in namespace factory.DelegateNameSpace are prefixed with 1-byte tag
+ * to indicate the type of stored object. This enables to quickly scan the
+ * entire bucket and filter out objects of a certain type
+ */
+const (
+	inActiveDel = byte(iota)
```
Comment (CoderZhi) [inline action/protocol/staking/delegate.go]: inactive

File: action/protocol/staking/delegate.go (hunk 1 above)
Comment (dustinxie) [reply action/protocol/staking/delegate.go]: to name staking protocol internal API

already added other APIs like stakingGet/PutBucket() etc.

File: action/protocol/staking/stakingpb/staking.proto (hunk 2 above)
Comment (dustinxie) [reply action/protocol/staking/stakingpb/staking.proto]: yes, but lots of proto used string to store, so just use string

File: action/protocol/staking/delegate.go (hunk 3 above)
Comment (dustinxie) [reply action/protocol/staking/delegate.go]: prefix for storing inactive and active delegates, see stakingPutDelegate()

so on Start(), can quickly scan the entire bucket and filter out active delegates

File: action/protocol/staking/delegate.go (hunk 4 above)
Comment (dustinxie) [reply action/protocol/staking/delegate.go]: chagned

File: action/protocol/staking/delegate.go (hunk 1 above)
Comment (CoderZhi) [reply action/protocol/staking/delegate.go]: as they are internal, the prefix is redundant.

**Pull request merged**